internal/querynode/segment_loader.go

```diff
@@ -139,24 +157,27 @@ func (loader *segmentLoader) loadSegment(req *querypb.LoadSegmentsRequest, segme
 		partitionID := loadInfo.PartitionID
 		segmentID := loadInfo.SegmentID
 		segment := newSegments[segmentID]
+
 		tr := timerecord.NewTimeRecorder("loadDurationPerSegment")
-		err = loader.loadSegmentInternal(segment, loadInfo)
+		err := loader.loadSegmentInternal(segment, loadInfo)
 		if err != nil {
 			log.Error("load segment failed when load data into memory",
 				zap.Int64("collectionID", collectionID),
 				zap.Int64("partitionID", partitionID),
 				zap.Int64("segmentID", segmentID),
-				zap.Int32("segment type", int32(segmentType)),
+				zap.Int32("segment type", int32(segment.getType())),
 				zap.Error(err))
 			return err
 		}
 
 		metrics.QueryNodeLoadSegmentLatency.WithLabelValues(fmt.Sprint(Params.QueryNodeCfg.QueryNodeID)).Observe(float64(tr.ElapseSpan().Milliseconds()))
-
 		return nil
 	}
 	// start to load
-	err = funcutil.ProcessFuncParallel(len(req.Infos), concurrencyLevel, loadSegmentFunc, "loadSegmentFunc")
+	// Make sure we can always benefit from concurrency, and not spawn too many idle goroutines
+	err = funcutil.ProcessFuncParallel(len(req.Infos),
```

Comment:
do we still need the concurrencyLevel or we just set it to segment numbers because the concurrency is already handled in loadSegmentInternal

Reply:
We still need it, concurrencyLevel controls how many segments are loaded concurrently, the inner pool controls how many fields are loaded concurrently. We need it to limit the memory usage during loading.

Reply:
I think we should have one centralized control for the overall concurrency.
We can issue segment load task as many as we want and left the concurrency control to the pool